src/librustc/ty/mod.rs

```diff
@@ -1726,6 +1728,190 @@ impl<'tcx, 'container> AdtDefData<'tcx, 'container> {
             None => NoDtor,
         }
     }
+
+    /// Returns a simpler type such that `Self: Sized` if and only
+    /// if that type is Sized, or `TyErr` if this type is recursive.
+    ///
+    /// HACK: instead of returning a list of types, this function can
+    /// return a tuple. In that case, the result is Sized only if
+    /// all elements of the tuple are Sized.
+    ///
+    /// This is generally the `struct_tail` if this is a struct, or a
+    /// tuple of them if this is an enum.
+    ///
+    /// Oddly enough, checking that the sized-constraint is Sized is
+    /// actually more expressive than checking all members:
+    /// the Sized trait is inductive, so an associated type that references
+    /// Self would prevent its containing ADT from being Sized.
+    ///
+    /// Due to normalization being eager, this applies even if
+    /// the associated type is behind a pointer, e.g. issue #31299.
+    pub fn sized_constraint(&self, tcx: &ty::TyCtxt<'tcx>) -> Ty<'tcx> {
+        let dep_node = DepNode::SizedConstraint(self.did);
+        match self.sized_constraint.get(dep_node) {
+            None => {
+                let this = tcx.lookup_adt_def_master(self.did);
+                this.calculate_sized_constraint_inner(tcx, &mut Vec::new());
+                self.sized_constraint(tcx)
+            }
+            Some(ty) => ty
+        }
+    }
+}
+
+impl<'tcx> AdtDefData<'tcx, 'tcx> {
+    fn sized_constraint_for_tys<TYS>(
+        &'tcx self,
+        tcx: &ty::TyCtxt<'tcx>,
+        stack: &mut Vec<AdtDefMaster<'tcx>>,
+        tys: TYS
+    ) -> Ty<'tcx>
+        where TYS: IntoIterator<Item=Ty<'tcx>>
+    {
+        let tys : Vec<_> = tys.into_iter()
+            .map(|ty| self.sized_constraint_for_ty(tcx, stack, ty))
+            .flat_map(|ty| match ty.sty {
+                ty::TyTuple(ref tys) => tys.last().cloned(),
+                _ => Some(ty)
+            })
+            .filter(|ty| *ty != tcx.types.bool)
+            .collect();
+
+        match tys.len() {
+            _ if tys.references_error() => tcx.types.err,
+            0 => tcx.types.bool,
+            1 => tys[0],
+            _ => tcx.mk_tup(tys)
+        }
+    }
+
+    fn sized_constraint_for_ty(
+        &'tcx self,
+        tcx: &ty::TyCtxt<'tcx>,
+        stack: &mut Vec<AdtDefMaster<'tcx>>,
+        ty: Ty<'tcx>
+    ) -> Ty<'tcx> {
+        let result = match ty.sty {
+            TyBool | TyChar | TyInt(..) | TyUint(..) | TyFloat(..) |
+            TyBox(..) | TyRawPtr(..) | TyRef(..) | TyFnDef(..) | TyFnPtr(_) |
+            TyArray(..) | TyClosure(..) => {
+                // these are always sized - return a primitive
```

Comment:
I assume part of the logic here is that, by always returning the *same* primitive, we help with caching, right? I think it's worth noting that explicitly in a comment.